oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/AbstractWebClientReactiveOAuth2AccessTokenResponseClient.java

```diff
@@ -97,7 +102,18 @@ private void populateTokenRequestHeaders(T grantRequest, HttpHeaders headers) {
 		headers.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON));
 		if (ClientAuthenticationMethod.CLIENT_SECRET_BASIC.equals(clientRegistration.getClientAuthenticationMethod())
 				|| ClientAuthenticationMethod.BASIC.equals(clientRegistration.getClientAuthenticationMethod())) {
-			headers.setBasicAuth(clientRegistration.getClientId(), clientRegistration.getClientSecret());
+			try {
+				String clientId = this.encodeClientCredentials
+						? URLEncoder.encode(clientRegistration.getClientId(), StandardCharsets.UTF_8.displayName())
+						: clientRegistration.getClientId();
+				String clientSecret = this.encodeClientCredentials
+						? URLEncoder.encode(clientRegistration.getClientSecret(), StandardCharsets.UTF_8.displayName())
+						: clientRegistration.getClientSecret();
+				headers.setBasicAuth(clientId, clientSecret);
+			}
+			catch (UnsupportedEncodingException ex) {
+				throw new IllegalStateException("Unable to URL-encode client id or client secret", ex);
```

Comment:
Let's throw `IllegalArgumentException` instead and ensure there is a test for this as well.